horovod/run/elastic/discovery.py

```diff
@@ -0,0 +1,139 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+from __future__ import absolute_import
+
+import logging
+import threading
+
+from collections import defaultdict
+
+import six
+
+from horovod.run.common.util import safe_shell_exec
+
+
+class HostState(object):
+    def __init__(self):
+        self._event = threading.Event()
+
+        # TODO(travis): blacklisted hosts should have a timeout period that increases with each failure
+        self._blacklisted = False
+
+    def get_event(self):
+        if self._event.is_set():
+            event = threading.Event()
+            self._event = event
+        return self._event
+
+    def set_event(self):
+        self._event.set()
+
+    def blacklist(self):
+        self._blacklisted = True
+        self.set_event()
+
+    def is_blacklisted(self):
+        return self._blacklisted
+
+
+class DiscoveredHosts(object):
+    def __init__(self, discovery):
+        self._available_hosts = set()
+        self._available_slots = {}
+        self._host_state = defaultdict(HostState)
+        self._discovery = discovery
+
+    def update_available_hosts(self):
+        prev_hosts = self._available_hosts
+        prev_slots = self._available_slots
+        available_hosts, available_slots = self._discovery.find_available_hosts_and_slots()
+        if prev_hosts != available_hosts or prev_slots != available_slots:
+            self._available_hosts, self._available_slots = available_hosts, available_slots
+            return True
+        return False
+
+    def count_available_slots(self):
+        return sum([self.get_slots(host) for host in self._available_hosts
+                    if not self._host_state[host].is_blacklisted()])
+
+    def get_slots(self, host):
+        return self._available_slots[host]
+
+    def filter_available_hosts(self, hosts):
+        return [host for host in hosts
+                if host in self._available_hosts and not self._host_state[host].is_blacklisted()]
+
+    def get_available_hosts(self):
+        return self.filter_available_hosts(self._available_hosts)
+
+    def blacklist(self, host):
+        if not self._host_state[host].is_blacklisted():
+            logging.warning('blacklist failing host: {}'.format(host))
+        self._host_state[host].blacklist()
+
+    def is_blacklisted(self, host):
+        return self._host_state[host].is_blacklisted()
+
+    def count_blacklisted_slots(self):
```

Comment:
I found that this method fails when hosts get removed from the `_available_slots` (because the discovery script does not list the host anymore). The host is still in `_host_state` but `get_slots(host)` fails with a `KeyError`.

Should this method call `get_slots(host)` only `if host in self._available_slots`?
Or should `get_slots(host)` return `0 if host not in self._available_slots`?

Would that change pose a problem for `WorkerStateRegistry._on_workers_recorded` where it says `# If there are no active hosts that aren't blacklisted, treat this as job failure` (which is btw. very difficult to understand)?

Reply:
Seems to be caused by a race condition.  In `count_blacklisted_slots`, we iterate over `self._available_hosts`, which should be 1:1 with `self._available_slots`, but it is possible that the discovery thread modifies this structure during the iteration.  

I propose two changes:

1. Use a default value if missing for `get_slots` of `0`.
2. Guard access to `available_hosts` and `available_slots` with a mutex.

I'll make this change.

Reply:
Okay, this turned into a larger refactor to ensure that the host update thread doesn't conflict with any operations on the set of known hosts.  This was accomplished by moving all operations to a separate class that is atomically updated.